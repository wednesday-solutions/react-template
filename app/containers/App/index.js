/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import React, { useEffect, useState } from 'react';
import { Switch, Route } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Router } from 'react-router';
import map from 'lodash/map';
import { Layout } from 'antd';
import { PersistGate } from 'redux-persist/integration/react';
import { routeConfig } from '@app/routeConfig';
import { ThemeProvider } from 'styled-components';
import GlobalStyle from '@app/global-styles';
import { colors } from '@themes';
import Header from '@components/Header';
import ScrollToTop from '@components/ScrollToTop';
import For from '@components/For';
import LanguageProvider from 'containers/LanguageProvider';
import ErrorBoundary from '@app/components/ErrorBoundary/index';
import { translationMessages } from '@app/i18n';
import { Provider } from 'react-redux';
import history from '@utils/history';
import configureStore from '@app/configureStore';
import If from '@app/components/If/index';

const theme = {
  fg: colors.primary,
  bg: colors.secondary
};

export function App() {
  const [store, setStore] = useState(null);
  const [persistor, setPersistor] = useState(null);

  const { location } = history;
  useEffect(() => {
    if (location.search.includes('?redirect_uri=')) {
      const routeToReplace = new URLSearchParams(location.search).get('redirect_uri');
      history.replace(routeToReplace);
    }
    const { store: s, persistor } = configureStore({}, history);
    setStore(s);
    setPersistor(persistor);
  }, []);

  return (
    <If condition={!!persistor} otherwise={<div>LOADING</div>}>
      <PersistGate loading={null} persistor={persistor}>
        <Router history={history}>
          <ScrollToTop>
            <ErrorBoundary>
              <Provider store={store}>
                <LanguageProvider messages={translationMessages}>
                  <ThemeProvider theme={theme}>
                    <Header />
                    <Layout.Content>
                      <For
                        ParentComponent={(props) => <Switch {...props} />}
                        of={map(Object.keys(routeConfig))}
                        renderItem={(routeKey, index) => {
                          const Component = routeConfig[routeKey].component;
                          return (
                            <Route
                              exact={routeConfig[routeKey].exact}
                              key={index}
                              path={routeConfig[routeKey].route}
                              render={(props) => {
                                const updatedProps = {
                                  ...props,
                                  ...routeConfig[routeKey].props
                                };
                                return <Component {...updatedProps} />;
                              }}
                            />
                          );
                        }}
                      />
                      <GlobalStyle />
                    </Layout.Content>
                  </ThemeProvider>
                </LanguageProvider>
              </Provider>
            </ErrorBoundary>
          </ScrollToTop>
        </Router>
      </PersistGate>
    </If>
  );
}
App.propTypes = {
  location: PropTypes.object,
  history: PropTypes.object
};
export default App;
