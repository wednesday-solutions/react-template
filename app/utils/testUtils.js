import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { browserHistory, BrowserRouter, Router } from 'react-router-dom';
import { ThemeProvider } from '@mui/material/styles';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import { IntlGlobalProvider } from '@components/IntlGlobalProvider';
import { theme } from '@containers/App';

export const renderWithIntl = (children) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <IntlGlobalProvider>{children}</IntlGlobalProvider>
    </IntlProvider>
  );

export const renderProvider = (children, history) => {
  const store = configureStore({}, browserHistory).store;
  return render(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider theme={theme}>
          {history ? <Router history={history}>{children}</Router> : <BrowserRouter>{children}</BrowserRouter>}
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});

export const getStyles = () => {};
