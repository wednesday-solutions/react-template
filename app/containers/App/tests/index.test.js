import React from 'react';
import { renderProvider } from '@utils/testUtils';
import App from '../index';
import { BrowserRouter } from 'react-router-dom';
// import history from '@app/utils/history';
// import { waitFor } from '@testing-library/react';

describe('<App /> container tests', () => {
  it('should render and match the snapshot', () => {
    const { container } = renderProvider(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  // it('should redirect to redirect_uri query params if given in url', async () => {
  //   history.location.search = '?redirect_uri=/repos';
  //   renderProvider(<App />, history);
  //   await waitFor(() => expect(history.location.pathname).toBe('/repos'));
  // });
});
