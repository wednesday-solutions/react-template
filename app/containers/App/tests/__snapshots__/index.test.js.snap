// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> container tests should render and match the snapshot 1`] = `<div />`;
